v2/src/log_stacktrace.c

```diff
@@ -0,0 +1,279 @@
+static void snprintf_fallback_impl(char** destination, size_t* destination_size, const char* fallback_string, size_t fallbackstring_size, const char* format, ...)
+{
+    va_list args;
+    va_start(args, format);
+
+    size_t copied;
+
+    int snprintfResult;
+
+    /*trying to complain*/
+    snprintfResult = vsnprintf(*destination, *destination_size, format, args);
+    if (!(
+        (snprintfResult >= 0) && /*the returned value is nonnegative [...]*/
+        (snprintfResult < (int)(*destination_size)) /*[...] and less than n.*/
+        ))
+    {
+        /*complain about not being able to complain*/
+        copied = memcat(*destination, *destination_size, fallback_string, fallbackstring_size - 1);
+        (*destination) += copied;
+        (*destination_size) -= copied;
+    }
+    else
+    {
+        /*complain*/
+        (*destination) += snprintfResult;
+        (*destination_size) -= snprintfResult;
+    }
+    va_end(args);
+}
+
+#if defined(_MSC_VER)
+/*this function only exists for Windows machines*/
+/*the function does the following
+1) parameter validation
+2) call SymInitialize for the current process
+3) get hThread's context.
+    3.1) if the current thread is hThread, call RtlCaptureContext
+    3.2) if the current thread is not hThread, call SuspendThread and GetThreadContext
+4) once the context has been acquired, call StackWalk64 to get the stack frames. For every frame:
+    4.1) determine the function name
+    4.2) determine the file name and line number
+    4.3) append the function name, file name and line number to the destination
+*/
+
+#if defined(_WIN64)
+    #define INSTRUCTION_POINTER_REGISTER Rip
+    #define FRAME_POINTER_REGISTER Rbp
+    #define STACK_POINTER_REGISTER Rsp
+#else
+    #if defined(_WIN32)
+        #define INSTRUCTION_POINTER_REGISTER Eip
+        #define FRAME_POINTER_REGISTER Ebp
+        #define STACK_POINTER_REGISTER Esp
+    #else
+        #error unknown version of windows
+    #endif
+#endif
+
+void getThreadStackAsString(HANDLE hThread, char* destination, size_t destinationSize)
+{
+
+    /*1) parameter validation*/
+    if ((destination == NULL) || (destinationSize == 0))
+    {
+        /*cannot compute if the output space is not sufficient (invalid args)*/
+        return;
+    }
+    else
+    {
+        destination[0] = '\0';
+
+        bool firstLine = true; /*only used to insert a \n between stack frames*/
+
+        HANDLE hProcess = GetCurrentProcess();
+
+        /*2) call SymInitialize for the current process*/
+        /*lazily call once SymInitialize*/
+        LONG state;
+        while ((state = InterlockedCompareExchange(&doSymInit, 1, 0)) != 2)
+        {
+            if (state == 0)
+            {
+                (void)SymInitialize(hProcess, NULL, TRUE); /*this is a process-wide initialization, and will leak because we don't know when to call SymCleanup. It is a good candidate for platform_init*/
```

Comment:
> It is a good candidate for platform_init*/ [](http://example.com/codeflow?start=160&length=43)

I think this module could benefit from having a log_stacktrace_init/deinit.
It could be added to logger_init/deinit, platform_init/deinit or left to the user event (although this would likely be sucky). Likely platform init would be a good candidate.

Reply:
For the current case that we have (where this "as written" works just nice) I don't see an added value to have this call to SymInitialize done in a separate fucntion.

However, we can do this with a different task, which I will put in "things that never get done 3".

Reply:
added [https://msazure.visualstudio.com/One/_workitems/edit/25733038](https://msazure.visualstudio.com/One/_workitems/edit/25733038)